astropy/units/decorators.py

```diff
@@ -5,10 +5,85 @@
 
 from ..utils.decorators import wraps
 from ..utils.compat import funcsigs
+from ..utils.misc import isiterable
 
 from .core import Unit, UnitsError, add_enabled_equivalencies
 from .physical import _unit_physical_mapping
 
+def _get_allowed_units(targets):
+    """
+    From a list of target units (either as strings or unit objects) and physical
+    types, return a list of Unit objects.
+    """
+
+    allowed_units = []
+    for target in targets:
+        if isinstance(target, str): # Could be a string unit or physical type
+
+            try: # unit passed in as a string
+                target_unit = Unit(target)
+
+            except ValueError:
+
+                try: # See if the function writer specified a physical type
+                    physical_type_id = _unit_physical_mapping[target]
+
+                except KeyError: # Function argument target is invalid
+                    raise ValueError("Invalid unit or physical type '{0}'."
+                                     .format(target))
+
+                # get unit directly from physical type id
+                target_unit = Unit._from_physical_type_id(physical_type_id)
+
+        else: # Remain agnostic, but it should be a Unit object if not a string
+            target_unit = target
+
+        allowed_units.append(target_unit)
+
+    return allowed_units
+
+def _validate_arg_value(param_name, func_name, arg, targets, equivalencies):
+    """
+    Validates the object passed in to the wrapped function, ``arg``, with target
+    unit or physical type, ``target``.
+    """
+
+    allowed_units = _get_allowed_units(targets)
+
+    # short-circuit if None is passed, and None is allowed
+    if arg is None and None in targets:
+        return
+
+    for allowed_unit in allowed_units:
+        try:
+            is_equivalent = arg.unit.is_equivalent(allowed_unit,
+                                                   equivalencies=equivalencies)
+
+            if is_equivalent:
+                break
+
+        except AttributeError: # Either there is no .unit or no .is_equivalent
+            if hasattr(arg, "unit"):
+                error_msg = "a 'unit' attribute without an 'is_equivalent' method"
+            else:
+                error_msg = "no 'unit' attribute"
+
+            raise TypeError("Argument '{0}' to function '{1}' has {2}. "
+                  "You may want to pass in an astropy Quantity instead."
+                     .format(param_name, func_name, error_msg))
+
+    else:
```

Comment:
Either this else statement or the above for statement has the incorrect indentation?